### PR TITLE
Custom non-blocking js alert

### PR DESCRIPTION
**Pull request by mallexxx**
Task/Issue URL: https://app.asana.com/0/715106103902962/1202323846903922/f
Tech Design URL:
CC: @miasma13 

**Description**:
Custom JS alert non-blocking browser interface an allowing to close the tab
https://www.figma.com/file/HZjSfXLUCIcTfDLFJAkM7h/JS-alerts?node-id=709%3A2601
- Requires product review

**Steps to test this PR**:
1. Validate alert/prompt/confirm js commands are working and returning valid result (https://w3cschools/jsref/met_win_alert.asp could be used for testing)
2. Validate tab can be closed and the browser UI is not blocked

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone X
* [ ] iPad

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

### CODE REVIEW

**Comment by Bunn** [comment]
@mallexxx 
while testing on my iPhone 13 mini - iOS 15.5 I can't manage to see the custom alerts.
It seems they are presented there, but the x/y position is wrong. You can see when I rotate the device it flashes for a split second.

https://user-images.githubusercontent.com/782316/179976888-5a0ca041-92c2-405d-bf01-ed0c06218c32.MP4

**Comment by Bunn** [comment]
@mallexxx I'm assigning this back to you, please assign it back to me once it's ready for review again. Thanks

File: Core/PixelEvent.swift [hunk 1]
```diff
@@ -482,6 +483,7 @@ extension Pixel.Event {
             
         case .jsAlertShown: return "m_js_alert_shown"
         case .jsAlertBlocked: return "m_js_alert_blocked"
+        case .jsAlertBackgroundTap: return "m_js_alert_bg_tap"
```
Comment (Bunn) [inline Core/PixelEvent.swift]: jsAlertBlocked is not being called, maybe it should be renamed to jsAlertTabClosed = "m_js_alert_tab_closed" since we don't have the block option anymore?

jsAlertBackgroundTap also not being called, when is this supposed to be called?

File: DuckDuckGo/TabViewController.swift [hunk 2]
```diff
@@ -419,11 +424,20 @@ class TabViewController: UIViewController {
     }
     
     public func load(url: URL) {
+        webView.stopLoading()
+        if jsAlertController.isShown {
+            jsAlertController.dismiss(animated: false)
+        }
+
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift]: This flag validation is being done in multiple places, what do you think about creating a func for it? Something like this:

```swift
  private func dismissJSAlertIfNecessary(animated: Bool) {
      if jsAlertController.isShown {
          jsAlertController.dismiss(animated: animated)
      }
  }

File: DuckDuckGo/TabViewController.swift [hunk 3]
```diff
@@ -419,11 +424,20 @@ class TabViewController: UIViewController {
...
         load(url: url, didUpgradeURL: false)
     }
     
     public func load(backForwardListItem: WKBackForwardListItem) {
         webView.stopLoading()
+        if jsAlertController.isShown {
+            jsAlertController.dismiss(animated: false)
+        }
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift]: Same as mentioned above about the flag validation.

File: DuckDuckGo/TabViewController.swift [hunk 4]
```diff
@@ -576,6 +590,10 @@ class TabViewController: UIViewController {
     }
     
     func goBack() {
+        if jsAlertController.isShown {
+            jsAlertController.dismiss(animated: false)
+        }
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift]: Same as mentioned above about the flag validation.

File: DuckDuckGo/TabViewController.swift [hunk 5]
```diff
@@ -590,6 +608,10 @@ class TabViewController: UIViewController {
     }
     
     func goForward() {
+        if jsAlertController.isShown {
+            jsAlertController.dismiss(animated: false)
+        }
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift]: Same as mentioned above about the flag validation.

File: DuckDuckGo/TabViewController.swift [hunk 6]
```diff
@@ -2002,6 +2045,10 @@ extension TabViewController: UIGestureRecognizerDelegate {
     }
 
     func refresh() {
+        if jsAlertController.isShown {
+            jsAlertController.dismiss(animated: false)
+        }
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift]: Same as mentioned above about the flag validation.

File: DuckDuckGo/JSAlertController.swift [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            self.textField.resignFirstResponder()
+        }
+        guard animated else {
+            self.alert = nil
+            self.view.superview?.isHidden = true
+            completion?()
+            return
+        }
+
+        UIView.animate(withDuration: 0.3) {
```
Comment (Bunn) [inline DuckDuckGo/JSAlertController.swift]: The present animation duration is 0.2 but the dismiss is 0.3 is this intentional?

File: DuckDuckGo/JSAlertController.swift [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        guard let isLocalUserInfoKey = notification.userInfo?[UIResponder.keyboardIsLocalUserInfoKey] as? NSNumber,
+              isLocalUserInfoKey == true,
+              let intersection = (notification.userInfo?[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect)?
+                  .intersection(self.view.convert(view.bounds, to: view.window)),
+              self.keyboardConstraint.constant != intersection.height
+        else {
+            return
+        }
+
+        UIView.animate(withDuration: 0.3) {
```
Comment (Bunn) [inline DuckDuckGo/JSAlertController.swift]: Maybe it would be good to create a private enum with all these animation durations?

**Review by Bunn — CHANGES_REQUESTED** [review]
@mallexxx Looks good, thanks for doing this :) 
I'm marking as Request Changes for 3 main reasons:
* TabViewController.swift, I think it would be good to concentrate that decision logic in a single function
* It would be good to call a pixel for when the tab is closed via jsAlertView to replace the previous pixel for blocked alerts
* We're missing a "Cancel" button on the text alert style

Before / After
before after


The other comments are small nitpicks.

Nice work!

File: DuckDuckGo/JSAlertController.swift (hunk 7 above)
Comment (mallexxx) [reply DuckDuckGo/JSAlertController.swift]: yes, this was intentional to match a system alert

**Comment by mallexxx** [comment]
Thanks @Bunn, this should be ready for review now

File: DuckDuckGo/TabViewController.swift [hunk 9]
```diff
@@ -71,38 +71,38 @@ class TabViewController: UIViewController {
...
+
     private var trackerNetworksDetectedOnPage = Set<String>()
     private var pageHasTrackers = false
-    
+
     private var detectedLoginURL: URL?
     private var preserveLoginsWorker: PreserveLoginsWorker?
-    
+
     private var trackersInfoWorkItem: DispatchWorkItem?
```
Comment (Bunn) [inline DuckDuckGo/TabViewController.swift:106]: Any reason for this? If not maybe would be better to discard this chunk. No strong opinions here, just trying to avoid unnecessary changes that seems to be cause by Xcode

File: DuckDuckGo/UserText.swift [hunk 10]
```diff
@@ -408,7 +408,9 @@ public struct UserText {
     
     public static let webJSAlertOKButton = NSLocalizedString("webJSAlert.OK.button", value: "OK", comment: "OK button for JavaScript alerts")
     public static let webJSAlertCancelButton = NSLocalizedString("webJSAlert.cancel.button", value: "Cancel", comment: "Cancel button for JavaScript alerts")
-        public static let noVoicePermissionAlertTitle = NSLocalizedString("voiceSearch.alert.no-permission.title", value: "Microphone Access Required", comment: "Title for alert warning the user about missing microphone permission")
+    public static let webJSAlertWebsiteMessageFormat = NSLocalizedString("webJSAlert.website-message.format", value: "A message from %@:", comment: "Alert title explaining the mssage is shown by a website")
```
Comment (Bunn) [inline DuckDuckGo/UserText.swift]: Typo -> `mssage`

File: DuckDuckGo/WebJSAlert.swift [hunk 11]
```diff
@@ -17,85 +17,66 @@
...
-struct WebJSAlert {
+class WebJSAlert {
+
     enum JSAlertType {
-        case confirm(handler: (_ blockAlerts: Bool, _ confirm: Bool) -> Void)
-        case text(handler: (_ blockAlerts: Bool, _ text: String?) -> Void, defaultText: String?)
-        case alert(handler: (_ blockAlerts: Bool) -> Void)
+        case confirm(handler: (_ confirm: Bool) -> Void, closeTab: () -> Void)
+        case text(handler: (_ text: String?) -> Void, defaultText: String?, closeTab: () -> Void)
+        case alert(handler: () -> Void, closeTab: () -> Void)
```
Comment (Bunn) [inline DuckDuckGo/WebJSAlert.swift]: I think we can remove references to closeTab since we're not using it anymore

File: DuckDuckGo/WebJSAlert.swift [hunk 12]
```diff
@@ -17,85 +17,66 @@
...
-                handler(true, nil)
-            }))
-            
-            return alertController
+        case .confirm(handler: _, closeTab: let closeTab):
+            closeTab()
+        case .text(handler: _, defaultText: _, closeTab: let closeTab):
+            closeTab()
+        case .alert(handler: _, closeTab: let closeTab):
+            closeTab()
```
Comment (Bunn) [inline DuckDuckGo/WebJSAlert.swift]: This can probably be removed

File: DuckDuckGo/JSAlertController.swift [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+        dismiss(animated: true) { [alert=self.alert] in
+            alert?.complete(with: false, text: nil)
+        }
+    }
+
+    @IBAction func closeTabAction(_ sender: Any) {
+        dismiss(animated: true) { [alert=self.alert] in
+            alert?.closeTab()
+        }
+    }
```
Comment (Bunn) [inline DuckDuckGo/JSAlertController.swift]: This can probably be removed.

**Review by Bunn — CHANGES_REQUESTED** [review]
@mallexxx The functionally is good, close tab button removed and cancel button added to the text alert.

I found a small typo and some old references to closeTab methods/closures that I don't think are necessary anymore. Other than that looks good.

File: DuckDuckGo/TabViewController.swift (hunk 9 above)
Comment (mallexxx) [reply DuckDuckGo/TabViewController.swift:106]: Xcode applies no leading spaces formatting to lines that were touched, I remember the onboarding task with code styling requirement to have this Xcode setting to be ON, I think we should put this on the Weekly Agenda and enforce this rule once again because I believe it‘s not strongly followed

**Comment by mallexxx** [comment]
Thanks once again :) Done

**Review by Bunn — APPROVED** [review]
LGTM ✅
Thanks :)
